packages/compiler/core/parser.ts

```diff
@@ -905,6 +907,23 @@ export function parse(code: string | SourceFile, options: ParseOptions = {}): Ca
     return expr;
   }
 
+  function parseArrayOrReferenceExpression(): ArrayExpressionNode | TypeReferenceNode {
```

Comment:
The name makes me feel like any array can be used, but it can only be an array of a reference expression if I reading the code right.

In other words, `model foo is {}[]` is not allowed, right.

Maybe `parseReferenceExpressionOrReferenceExpressionArray`

Mouthful, but it's clearer.